{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2c6b5224",
   "metadata": {},
   "source": [
    " Preprocess and convert single-celldata in anndata (h5ad) format to loom for input to Geneformer\n",
    " - Uses same gene IDs that Geneformer was trained on\n",
    " - Need to provide .h5ad dataset with labels for finetuning\n",
    " - The preprocessing here is minimal and it is recommended to do other processing possibly including thresholding on reads, removing irrelevant cells types, or other steps specific to your application area\n",
    " - Change pathes/file names in Parameters section.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e23de9-33bd-477e-a4a5-fef0bf4970c4",
   "metadata": {},
   "source": [
    "# Imports\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2f74de4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "import anndata as ad\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ddfb59",
   "metadata": {},
   "source": [
    "# Download and Read input data to finetune the Geneformer pre-trained model\n",
    "\n",
    "Many repositories have scRNAseq data in h5ad format, including:\n",
    " \n",
    "   - [scPerturb](https://zenodo.org/record/7041849) Single-Cell Perturbation Database.\n",
    "   \n",
    "        example data downlaod from scPerturb: `!wget -P {input_data_filepath} https://zenodo.org/record/7041849/files/AissaBenevolenskaya2021.h5ad`\n",
    "     \n",
    "     \n",
    "  -  [Cellxgene](https://cellxgene.cziscience.com/) Data Portal-  For example, the following is the paper and its corresponding data on Cellxgene. \n",
    "  \n",
    "     - [Paper (Chen et al. 2021)](https://www.cell.com/cell/fulltext/S0092-8674(21)01381-7)\n",
    "       \n",
    "     - [Link to the data on Cellxgene](https://cellxgene.cziscience.com/collections/a48f5033-3438-4550-8574-cdff3263fdfd)\n",
    "     \n",
    "  **Note:** the URL is valid for one week. If it expires, get a fresh link from the Cellxgene data portal.\n",
    "  \n",
    "  -  Other publically available resources\n",
    "  \n",
    "          - Here, we use a small public data as a finetuning example. Other desired single-cell data in h5ad format can replace it.   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46859f59",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ee2bf53",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_data_filepath = \"/home/domino/geneformer_workflow/input/data/publicdata\"\n",
    "h5ad_filename = \"adata_SS2_for_download.h5ad\"#\"GehringPachter2019.h5ad\"#\"AissaBenevolenskaya2021.h5ad\"#\"Chen_2021_VAL_DIS_Non_Epithelial.h5ad\"\n",
    "\n",
    "id_map_dir = \"/home/domino/geneformer_workflow/input/id_map\"\n",
    "\n",
    "results_dir = \"/home/domino/geneformer_workflow/results/perturbation\"\n",
    "os.makedirs(results_dir, exist_ok=True)\n",
    "\n",
    "loom_dir = os.path.join(results_dir, \"loom_files\")\n",
    "os.makedirs(loom_dir, exist_ok=True)\n",
    "loom_filepath = os.path.join(loom_dir, \"raw.loom\")\n",
    "\n",
    "adata_filepath = os.path.join(results_dir, 'adata.pk')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8f3f01b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column name in adata for label to be used by geneformer finetuning\n",
    "label_colname = \"celltype\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "274fd01e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(os.path.join(input_data_filepath, h5ad_filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "da268cc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1349 × 21030\n",
       "    obs: 'batchcode', 'treatment', 'celltype', 'n_genes', 'lane', 'percent_mito', 'n_counts', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch'\n",
       "    var: 'n_cells-0', 'n_cells-1', 'n_cells-2', 'n_cells-3'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fc3ebc10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batchcode</th>\n",
       "      <th>treatment</th>\n",
       "      <th>celltype</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>lane</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>scrublet_cluster_score</th>\n",
       "      <th>bh_pval</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C02</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>1264</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.039233</td>\n",
       "      <td>18199.0</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2499_C03</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>877</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.080071</td>\n",
       "      <td>14000.0</td>\n",
       "      <td>0.028409</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C03</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>1786</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.054565</td>\n",
       "      <td>131017.0</td>\n",
       "      <td>0.257732</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.808411</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2499_C04</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>394</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.081609</td>\n",
       "      <td>14496.0</td>\n",
       "      <td>0.064000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C04</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>544</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.099694</td>\n",
       "      <td>14364.0</td>\n",
       "      <td>0.109890</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2513_B10</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>SM</td>\n",
       "      <td>2177</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.053583</td>\n",
       "      <td>385477.0</td>\n",
       "      <td>0.175573</td>\n",
       "      <td>0.234129</td>\n",
       "      <td>0.755951</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2510_B11</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>SM</td>\n",
       "      <td>1805</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.056122</td>\n",
       "      <td>229909.0</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.234129</td>\n",
       "      <td>0.755951</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_C01</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>N</td>\n",
       "      <td>303</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.072749</td>\n",
       "      <td>1677.0</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_C02</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>N</td>\n",
       "      <td>521</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.085348</td>\n",
       "      <td>3351.0</td>\n",
       "      <td>0.032836</td>\n",
       "      <td>0.064000</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_A05</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>USM</td>\n",
       "      <td>1768</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.058195</td>\n",
       "      <td>94218.0</td>\n",
       "      <td>0.148649</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1349 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               batchcode treatment celltype  n_genes     lane  percent_mito  \\\n",
       "SCGC--2500_C02    BATCH3   CONTROL        N     1264  24057_6      0.039233   \n",
       "SCGC--2499_C03    BATCH3   CONTROL        N      877  24057_6      0.080071   \n",
       "SCGC--2500_C03    BATCH3   CONTROL        N     1786  24057_6      0.054565   \n",
       "SCGC--2499_C04    BATCH3   CONTROL        N      394  24057_6      0.081609   \n",
       "SCGC--2500_C04    BATCH3   CONTROL        N      544  24057_6      0.099694   \n",
       "...                  ...       ...      ...      ...      ...           ...   \n",
       "SCGC--2513_B10    BATCH1      CVID       SM     2177  24130_7      0.053583   \n",
       "SCGC--2510_B11    BATCH1   CONTROL       SM     1805  24130_7      0.056122   \n",
       "SCGC--2502_C01    BATCH1      CVID        N      303  24130_7      0.072749   \n",
       "SCGC--2502_C02    BATCH1      CVID        N      521  24130_7      0.085348   \n",
       "SCGC--2502_A05    BATCH1      CVID      USM     1768  24130_7      0.058195   \n",
       "\n",
       "                n_counts  scrublet_score  scrublet_cluster_score   bh_pval  \\\n",
       "SCGC--2500_C02   18199.0        0.095477                0.095477  0.975326   \n",
       "SCGC--2499_C03   14000.0        0.028409                0.095477  0.975326   \n",
       "SCGC--2500_C03  131017.0        0.257732                0.210526  0.808411   \n",
       "SCGC--2499_C04   14496.0        0.064000                0.083333  0.975326   \n",
       "SCGC--2500_C04   14364.0        0.109890                0.095477  0.975326   \n",
       "...                  ...             ...                     ...       ...   \n",
       "SCGC--2513_B10  385477.0        0.175573                0.234129  0.755951   \n",
       "SCGC--2510_B11  229909.0        0.210526                0.234129  0.755951   \n",
       "SCGC--2502_C01    1677.0        0.056180                0.056180  0.915471   \n",
       "SCGC--2502_C02    3351.0        0.032836                0.064000  0.915471   \n",
       "SCGC--2502_A05   94218.0        0.148649                0.056180  0.915471   \n",
       "\n",
       "               batch  \n",
       "SCGC--2500_C02     0  \n",
       "SCGC--2499_C03     0  \n",
       "SCGC--2500_C03     0  \n",
       "SCGC--2499_C04     0  \n",
       "SCGC--2500_C04     0  \n",
       "...              ...  \n",
       "SCGC--2513_B10     3  \n",
       "SCGC--2510_B11     3  \n",
       "SCGC--2502_C01     3  \n",
       "SCGC--2502_C02     3  \n",
       "SCGC--2502_A05     3  \n",
       "\n",
       "[1349 rows x 11 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0a1973",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells-2</th>\n",
       "      <th>n_cells-3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>53.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>64.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>77.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>81.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-21264C1.2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-2189O9.3</th>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21030 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_cells-0  n_cells-1  n_cells-2  n_cells-3\n",
       "A1BG               53.0       57.0       36.0       50.0\n",
       "A1BG-AS1           15.0       13.0        7.0        8.0\n",
       "A1CF                5.0        8.0        NaN        9.0\n",
       "A2M                 9.0       12.0        7.0        8.0\n",
       "A2ML1               5.0        6.0        3.0       10.0\n",
       "...                 ...        ...        ...        ...\n",
       "ZYX                64.0       63.0       53.0       61.0\n",
       "ZZEF1              77.0       94.0       72.0       79.0\n",
       "ZZZ3               81.0       77.0       58.0       80.0\n",
       "bP-21264C1.2        7.0       10.0        4.0        7.0\n",
       "bP-2189O9.3         6.0        4.0        NaN        3.0\n",
       "\n",
       "[21030 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1afdae4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-10 14:57:29--  https://huggingface.co/datasets/ctheodoris/Genecorpus-30M/resolve/main/example_input_files/gene_info_table.csv\n",
      "Resolving proxy-server.bms.com (proxy-server.bms.com)... 165.89.114.150\n",
      "Connecting to proxy-server.bms.com (proxy-server.bms.com)|165.89.114.150|:8080... connected.\n",
      "Proxy request sent, awaiting response... 200 OK\n",
      "Length: 3460504 (3.3M) [text/plain]\n",
      "Saving to: ‘/home/domino/geneformer_workflow/input/id_map/gene_info_table.csv’\n",
      "\n",
      "gene_info_table.csv 100%[===================>]   3.30M  8.49MB/s    in 0.4s    \n",
      "\n",
      "2023-10-10 14:57:29 (8.49 MB/s) - ‘/home/domino/geneformer_workflow/input/id_map/gene_info_table.csv’ saved [3460504/3460504]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -P {id_map_dir} https://huggingface.co/datasets/ctheodoris/Genecorpus-30M/resolve/main/example_input_files/gene_info_table.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9b1ed7ae",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80524</th>\n",
       "      <td>80524</td>\n",
       "      <td>ENSGunknown_pk</td>\n",
       "      <td>pk</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80525</th>\n",
       "      <td>80525</td>\n",
       "      <td>ENSGunknown_uc-338</td>\n",
       "      <td>uc-338</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80526</th>\n",
       "      <td>80526</td>\n",
       "      <td>ENSGunknown_uc-338_1</td>\n",
       "      <td>uc-338.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80527</th>\n",
       "      <td>80527</td>\n",
       "      <td>ENSGunknown_uc-338_2</td>\n",
       "      <td>uc-338.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80528</th>\n",
       "      <td>80528</td>\n",
       "      <td>ENSGunknown_uc-338_3</td>\n",
       "      <td>uc-338.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80529 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0            ensembl_id gene_name       gene_type\n",
       "0               0       ENSG00000000003    TSPAN6  protein_coding\n",
       "1               1       ENSG00000000005      TNMD  protein_coding\n",
       "2               2       ENSG00000000419      DPM1  protein_coding\n",
       "3               3       ENSG00000000457     SCYL3  protein_coding\n",
       "4               4       ENSG00000000460  C1orf112  protein_coding\n",
       "...           ...                   ...       ...             ...\n",
       "80524       80524        ENSGunknown_pk        pk             NaN\n",
       "80525       80525    ENSGunknown_uc-338    uc-338             NaN\n",
       "80526       80526  ENSGunknown_uc-338_1  uc-338.1             NaN\n",
       "80527       80527  ENSGunknown_uc-338_2  uc-338.2             NaN\n",
       "80528       80528  ENSGunknown_uc-338_3  uc-338.3             NaN\n",
       "\n",
       "[80529 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_csv(os.path.join(id_map_dir, \"gene_info_table.csv\"))\n",
    "gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbeefae4",
   "metadata": {},
   "source": [
    "# Process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8e505cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove gene names with multiple ensembl mappings\n",
    "gene_info = gene_info.drop_duplicates(subset='gene_name', keep=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e1714b01",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['gene_name'] = adata.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "40cd8991",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells-2</th>\n",
       "      <th>n_cells-3</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>53.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>A1CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>A2ML1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>64.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>77.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>ZZEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>81.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-21264C1.2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>bP-21264C1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-2189O9.3</th>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>bP-2189O9.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21030 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_cells-0  n_cells-1  n_cells-2  n_cells-3     gene_name\n",
       "A1BG               53.0       57.0       36.0       50.0          A1BG\n",
       "A1BG-AS1           15.0       13.0        7.0        8.0      A1BG-AS1\n",
       "A1CF                5.0        8.0        NaN        9.0          A1CF\n",
       "A2M                 9.0       12.0        7.0        8.0           A2M\n",
       "A2ML1               5.0        6.0        3.0       10.0         A2ML1\n",
       "...                 ...        ...        ...        ...           ...\n",
       "ZYX                64.0       63.0       53.0       61.0           ZYX\n",
       "ZZEF1              77.0       94.0       72.0       79.0         ZZEF1\n",
       "ZZZ3               81.0       77.0       58.0       80.0          ZZZ3\n",
       "bP-21264C1.2        7.0       10.0        4.0        7.0  bP-21264C1.2\n",
       "bP-2189O9.3         6.0        4.0        NaN        3.0   bP-2189O9.3\n",
       "\n",
       "[21030 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "48b61362",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = adata.var.merge(gene_info[['gene_name', 'ensembl_id']],\n",
    "                            on='gene_name',\n",
    "                            how='left').set_index('gene_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6cdbc47b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells-2</th>\n",
       "      <th>n_cells-3</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>53.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>64.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>77.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>81.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-21264C1.2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSGunknown_bP-21264C1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-2189O9.3</th>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSGunknown_bP-2189O9_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21030 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_cells-0  n_cells-1  n_cells-2  n_cells-3  \\\n",
       "gene_name                                                  \n",
       "A1BG               53.0       57.0       36.0       50.0   \n",
       "A1BG-AS1           15.0       13.0        7.0        8.0   \n",
       "A1CF                5.0        8.0        NaN        9.0   \n",
       "A2M                 9.0       12.0        7.0        8.0   \n",
       "A2ML1               5.0        6.0        3.0       10.0   \n",
       "...                 ...        ...        ...        ...   \n",
       "ZYX                64.0       63.0       53.0       61.0   \n",
       "ZZEF1              77.0       94.0       72.0       79.0   \n",
       "ZZZ3               81.0       77.0       58.0       80.0   \n",
       "bP-21264C1.2        7.0       10.0        4.0        7.0   \n",
       "bP-2189O9.3         6.0        4.0        NaN        3.0   \n",
       "\n",
       "                            ensembl_id  \n",
       "gene_name                               \n",
       "A1BG                   ENSG00000121410  \n",
       "A1BG-AS1               ENSG00000268895  \n",
       "A1CF                   ENSG00000148584  \n",
       "A2M                    ENSG00000175899  \n",
       "A2ML1                  ENSG00000166535  \n",
       "...                                ...  \n",
       "ZYX                    ENSG00000159840  \n",
       "ZZEF1                  ENSG00000074755  \n",
       "ZZZ3                   ENSG00000036549  \n",
       "bP-21264C1.2  ENSGunknown_bP-21264C1_2  \n",
       "bP-2189O9.3    ENSGunknown_bP-2189O9_3  \n",
       "\n",
       "[21030 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8e0d8f44",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/tbio/domino_202309/binaries/python-3.10.6/lib/python3.10/site-packages/anndata/_core/anndata.py:1113: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if not is_categorical_dtype(df_full[k]):\n"
     ]
    }
   ],
   "source": [
    "# Remove genes that didn't map to ensembl id\n",
    "adata = adata[:, ~adata.var['ensembl_id'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "433e63a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>n_cells-2</th>\n",
       "      <th>n_cells-3</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>53.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>64.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>77.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>81.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-21264C1.2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSGunknown_bP-21264C1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bP-2189O9.3</th>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSGunknown_bP-2189O9_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19298 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_cells-0  n_cells-1  n_cells-2  n_cells-3  \\\n",
       "gene_name                                                  \n",
       "A1BG               53.0       57.0       36.0       50.0   \n",
       "A1BG-AS1           15.0       13.0        7.0        8.0   \n",
       "A1CF                5.0        8.0        NaN        9.0   \n",
       "A2M                 9.0       12.0        7.0        8.0   \n",
       "A2ML1               5.0        6.0        3.0       10.0   \n",
       "...                 ...        ...        ...        ...   \n",
       "ZYX                64.0       63.0       53.0       61.0   \n",
       "ZZEF1              77.0       94.0       72.0       79.0   \n",
       "ZZZ3               81.0       77.0       58.0       80.0   \n",
       "bP-21264C1.2        7.0       10.0        4.0        7.0   \n",
       "bP-2189O9.3         6.0        4.0        NaN        3.0   \n",
       "\n",
       "                            ensembl_id  \n",
       "gene_name                               \n",
       "A1BG                   ENSG00000121410  \n",
       "A1BG-AS1               ENSG00000268895  \n",
       "A1CF                   ENSG00000148584  \n",
       "A2M                    ENSG00000175899  \n",
       "A2ML1                  ENSG00000166535  \n",
       "...                                ...  \n",
       "ZYX                    ENSG00000159840  \n",
       "ZZEF1                  ENSG00000074755  \n",
       "ZZZ3                   ENSG00000036549  \n",
       "bP-21264C1.2  ENSGunknown_bP-21264C1_2  \n",
       "bP-2189O9.3    ENSGunknown_bP-2189O9_3  \n",
       "\n",
       "[19298 rows x 5 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4907bf9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/tbio/domino_202309/binaries/python-3.10.6/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n",
      "/opt/tbio/domino_202309/binaries/python-3.10.6/lib/python3.10/site-packages/anndata/_core/anndata.py:1113: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if not is_categorical_dtype(df_full[k]):\n",
      "/opt/tbio/domino_202309/binaries/python-3.10.6/lib/python3.10/site-packages/anndata/_core/anndata.py:1113: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if not is_categorical_dtype(df_full[k]):\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ca9384fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batchcode</th>\n",
       "      <th>treatment</th>\n",
       "      <th>celltype</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>lane</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>scrublet_cluster_score</th>\n",
       "      <th>bh_pval</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C02</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>1161</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.039233</td>\n",
       "      <td>16209</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2499_C03</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>806</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.080071</td>\n",
       "      <td>12543</td>\n",
       "      <td>0.028409</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C03</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>1651</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.054565</td>\n",
       "      <td>115491</td>\n",
       "      <td>0.257732</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.808411</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2499_C04</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>347</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.081609</td>\n",
       "      <td>12357</td>\n",
       "      <td>0.064000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2500_C04</th>\n",
       "      <td>BATCH3</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>N</td>\n",
       "      <td>502</td>\n",
       "      <td>24057_6</td>\n",
       "      <td>0.099694</td>\n",
       "      <td>12333</td>\n",
       "      <td>0.109890</td>\n",
       "      <td>0.095477</td>\n",
       "      <td>0.975326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2513_B10</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>SM</td>\n",
       "      <td>1983</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.053583</td>\n",
       "      <td>330785</td>\n",
       "      <td>0.175573</td>\n",
       "      <td>0.234129</td>\n",
       "      <td>0.755951</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2510_B11</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>SM</td>\n",
       "      <td>1673</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.056122</td>\n",
       "      <td>204929</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.234129</td>\n",
       "      <td>0.755951</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_C01</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>N</td>\n",
       "      <td>289</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.072749</td>\n",
       "      <td>1516</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_C02</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>N</td>\n",
       "      <td>486</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.085348</td>\n",
       "      <td>3097</td>\n",
       "      <td>0.032836</td>\n",
       "      <td>0.064000</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGC--2502_A05</th>\n",
       "      <td>BATCH1</td>\n",
       "      <td>CVID</td>\n",
       "      <td>USM</td>\n",
       "      <td>1624</td>\n",
       "      <td>24130_7</td>\n",
       "      <td>0.058195</td>\n",
       "      <td>82912</td>\n",
       "      <td>0.148649</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.915471</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1333 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               batchcode treatment celltype  n_genes     lane  percent_mito  \\\n",
       "SCGC--2500_C02    BATCH3   CONTROL        N     1161  24057_6      0.039233   \n",
       "SCGC--2499_C03    BATCH3   CONTROL        N      806  24057_6      0.080071   \n",
       "SCGC--2500_C03    BATCH3   CONTROL        N     1651  24057_6      0.054565   \n",
       "SCGC--2499_C04    BATCH3   CONTROL        N      347  24057_6      0.081609   \n",
       "SCGC--2500_C04    BATCH3   CONTROL        N      502  24057_6      0.099694   \n",
       "...                  ...       ...      ...      ...      ...           ...   \n",
       "SCGC--2513_B10    BATCH1      CVID       SM     1983  24130_7      0.053583   \n",
       "SCGC--2510_B11    BATCH1   CONTROL       SM     1673  24130_7      0.056122   \n",
       "SCGC--2502_C01    BATCH1      CVID        N      289  24130_7      0.072749   \n",
       "SCGC--2502_C02    BATCH1      CVID        N      486  24130_7      0.085348   \n",
       "SCGC--2502_A05    BATCH1      CVID      USM     1624  24130_7      0.058195   \n",
       "\n",
       "                n_counts  scrublet_score  scrublet_cluster_score   bh_pval  \\\n",
       "SCGC--2500_C02     16209        0.095477                0.095477  0.975326   \n",
       "SCGC--2499_C03     12543        0.028409                0.095477  0.975326   \n",
       "SCGC--2500_C03    115491        0.257732                0.210526  0.808411   \n",
       "SCGC--2499_C04     12357        0.064000                0.083333  0.975326   \n",
       "SCGC--2500_C04     12333        0.109890                0.095477  0.975326   \n",
       "...                  ...             ...                     ...       ...   \n",
       "SCGC--2513_B10    330785        0.175573                0.234129  0.755951   \n",
       "SCGC--2510_B11    204929        0.210526                0.234129  0.755951   \n",
       "SCGC--2502_C01      1516        0.056180                0.056180  0.915471   \n",
       "SCGC--2502_C02      3097        0.032836                0.064000  0.915471   \n",
       "SCGC--2502_A05     82912        0.148649                0.056180  0.915471   \n",
       "\n",
       "               batch  \n",
       "SCGC--2500_C02     0  \n",
       "SCGC--2499_C03     0  \n",
       "SCGC--2500_C03     0  \n",
       "SCGC--2499_C04     0  \n",
       "SCGC--2500_C04     0  \n",
       "...              ...  \n",
       "SCGC--2513_B10     3  \n",
       "SCGC--2510_B11     3  \n",
       "SCGC--2502_C01     3  \n",
       "SCGC--2502_C02     3  \n",
       "SCGC--2502_A05     3  \n",
       "\n",
       "[1333 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cells should be labeled with the total read count in the cell (column attribute \"n_counts\")\n",
    "# to be used for normalization by geneformer.\n",
    "adata.obs[\"n_counts\"] = np.squeeze(np.asarray(np.sum(adata.X, axis=1))).astype(\"int\")\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bbb6867b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1333 × 19288\n",
       "    obs: 'batchcode', 'treatment', 'celltype', 'n_genes', 'lane', 'percent_mito', 'n_counts', 'scrublet_score', 'scrublet_cluster_score', 'bh_pval', 'batch'\n",
       "    var: 'n_cells-0', 'n_cells-1', 'n_cells-2', 'n_cells-3', 'ensembl_id', 'n_cells'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edb073e",
   "metadata": {},
   "source": [
    "# Write output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e564c034",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write adata object\n",
    "with open(adata_filepath, 'wb') as outfile:\n",
    "    pickle.dump(adata, outfile, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a9d34a58",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/domino/.local/lib/python3.10/site-packages/loompy/bus_file.py:68: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def twobit_to_dna(twobit: int, size: int) -> str:\n",
      "/home/domino/.local/lib/python3.10/site-packages/loompy/bus_file.py:85: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def dna_to_twobit(dna: str) -> int:\n",
      "/home/domino/.local/lib/python3.10/site-packages/loompy/bus_file.py:102: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  def twobit_1hamming(twobit: int, size: int) -> List[int]:\n"
     ]
    }
   ],
   "source": [
    "# Write out as loom as well\n",
    "adata.write_loom(loom_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16aa002c-086f-4982-b9f3-8fda97dbc30e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
